source/fe/fe_rannacher_turek.cc

```diff
@@ -0,0 +1,119 @@
+#include <deal.II/base/quadrature_lib.h>
+#include <algorithm>
+
+#include <sstream>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+template <int dim>
+FE_RannacherTurek<dim>::FE_RannacherTurek(const unsigned int degree,
+                                          const unsigned int n_face_support_points) :
+  FE_Poly<PolynomialsRannacherTurek<dim>, dim>(
+    PolynomialsRannacherTurek<dim>(),
+    FiniteElementData<dim>(this->get_dpo_vector(),
+                           1,
+                           2,
+                           FiniteElementData<dim>::L2),
+    std::vector<bool>(4, false), // restriction not implemented
+    std::vector<ComponentMask>(4, std::vector<bool>(1, true))),
+  degree(degree),
+  n_face_support_points(n_face_support_points)
+{
+  Assert(dim == 2, ExcNotImplemented());
+  Assert(degree == 0, ExcNotImplemented());
+  this->initialize_support_points();
+}
+
+
+
+template <int dim>
+std::vector<unsigned int> FE_RannacherTurek<dim>::get_dpo_vector()
+{
+  std::vector<unsigned int> dpo(dim + 1, 0);
+  dpo[dim - 1] = 1;
+
+  return dpo;
+}
+
+
+
+template <int dim>
+std::string FE_RannacherTurek<dim>::get_name() const
+{
+  std::ostringstream namebuf;
+  namebuf << "FE_RannacherTurek"
+          << "<" << dim << ">"
+          << "(" << this->degree << ", " << this-n_face_support_points << ")";
+  return namebuf.str();
+}
+
+
+
+template <int dim>
+FiniteElement<dim> *FE_RannacherTurek<dim>::clone() const
+{
+  return new FE_RannacherTurek<dim>(this->degree, this->n_face_support_points);
+}
+
+
+
+template <int dim>
+void FE_RannacherTurek<dim>::initialize_support_points()
+{
+  Assert(dim == 2, ExcNotImplemented());
+  dealii::QGauss<dim-1> face_quadrature(this->n_face_support_points);
+  this->weights = face_quadrature.get_weights();
+  this->generalized_support_points.resize(4*face_quadrature.size());
+  for (std::size_t q = 0;
+       q < face_quadrature.size();
+       ++q)
+    {
+      this->generalized_support_points[0*face_quadrature.size() + q] =
+        dealii::Point<dim>(0, 1 - face_quadrature.point(q)(0));
+      this->generalized_support_points[1*face_quadrature.size() + q] =
+        dealii::Point<dim>(1, 1 - face_quadrature.point(q)(0));
+      this->generalized_support_points[2*face_quadrature.size() + q] =
+        dealii::Point<dim>(face_quadrature.point(q)(0), 0);
+      this->generalized_support_points[3*face_quadrature.size() + q] =
+        dealii::Point<dim>(face_quadrature.point(q)(0), 1);
+    }
+}
+
+
+
+template <int dim>
+void FE_RannacherTurek<dim>::interpolate(
+  std::vector<double> &local_dofs,
+  const std::vector<double> &values) const
+{
+  AssertDimension(values.size(), this->generalized_support_points.size());
+  AssertDimension(local_dofs.size(), this->dofs_per_cell);
+
+  std::size_t q_points_per_face = this->weights.size();
+  std::fill(local_dofs.begin(), local_dofs.end(), 0.0);
+
+  typedef std::vector<double>::const_iterator it;
+  it value = values.begin();
```

Comment:
You only use the typedef once. Let's just spell things out to make code easier to read.